File: src/main/java/org/excavator/grpc/GrpcServerApplication.java
package org.excavator.grpc;

import io.grpc.Server;
import io.grpc.ServerBuilder;

import java.io.IOException;

public class GrpcServerApplication {

    public static void main(String[] args) throws InterruptedException, IOException {
        Server server = ServerBuilder.forPort(53000)
                .addService(new ProductService())
                .build()
                .start();

        Runtime.getRuntime().addShutdownHook(new Thread(server::shutdownNow));
        server.awaitTermination();
    }
}

File: src/main/scala/org/excavator/grpc/GrpcServerApplication.scala
package org.excavator.grpc

import java.util.concurrent.TimeUnit

import io.grpc.{Server, ServerBuilder}

class GrpcServerApplication {

  def start(port: Int) = {

    val p = new ProductServiceRpc()

    val server = ServerBuilder.forPort(port)
      .addService(p)
      .build()
      .start()

    addHook(server)
    server.awaitTermination(10, TimeUnit.SECONDS)
  }

  private def addHook(server: Server) = {
    Runtime.getRuntime.addShutdownHook(new Thread(){
      override def run(): Unit = {
        server.shutdownNow()
      }
    })
  }

}
